src/workerd/io/trace.h

```diff
@@ -519,4 +309,182 @@ inline kj::String truncateScriptId(kj::StringPtr id) {
+  // and reports spans gets the `Span` type.
+  //
+  // The reason we use a separate builder type rather than rely on constness is so that the methods
+  // can be no-ops when there is no observer, avoiding unnecessary allocations. To allow for this,
+  // SpanBuilder is designed to be write-only -- you cannot read back the content. Only the
+  // observer (if their is one) receives the content.
+
+public:
+  explicit SpanBuilder(SpanParent& parent, kj::StringPtr operationName)
+      : SpanBuilder(parent.newChild(), operationName) {}
+  explicit SpanBuilder(SpanParent&& parent, kj::StringPtr operationName)
+      : SpanBuilder(parent.newChild(), operationName) {}
+  // Create a new child span from the given parent.
+  //
+  // `operationName` should be a string literal with infinite lifetime.
+
+  explicit SpanBuilder(kj::Maybe<kj::Own<SpanObserver>> observer, kj::StringPtr operationName)
+      { KJ_IF_MAYBE(o, observer) { state.emplace(kj::mv(*o), operationName); } }
+  // Create a new top-level span that will report to the given observer. If the observer is null,
+  // no data is collected.
+  //
+  // `operationName` should be a string literal with infinite lifetime.
+
+  ~SpanBuilder() noexcept(false);
+
+  void setStartTime(kj::Date time);
+  // Set a different start time. This is optional; by default the time when the builder was
+  // constructed will be used.
+
+  using TagValue = Span::TagValue;
+  void setTag(kj::StringPtr key, TagValue value);
+  // `key` must point to memory that will remain valid all the way until this span's data is
+  // serialized.
+
+  void addLog(kj::Date timestamp, kj::StringPtr key, TagValue value);
+  // `key` must point to memory that will remain valid all the way until this span's data is
+  // serialized.
+  //
+  // The differences between this and `setTag()` is that logs are timestamped and may have
+  // duplicate keys.
+
+private:
+  struct State {
+    kj::Own<SpanObserver> observer;
+    // The observer, or null if not being observed.
+
+    Span span;
+    // The under-construction span.
+
+    State(kj::Own<SpanObserver> observer, kj::StringPtr operationName)
+        : observer(kj::mv(observer)), span(operationName) {}
+  };
+  kj::Maybe<State> state;
+
+  friend class SpanParent;
+};
+
+class SpanObserver: public kj::Refcounted {
+  // Abstract interface for observing trace spans reported by the runtime. Different
+  // implementations might support different tracing back-ends, e.g. Trace Workers, Jaeger, or
+  // whatever infrastrure you prefer to use for this.
+  //
+  // A new SpanObserver is created at the start of each Span. The observer is used to report the
+  // span data at the end of the span, as well as to construct child observers.
+
+public:
+  virtual kj::Own<SpanObserver> newChild() = 0;
+  // Allocate a new child span.
+  //
+  // Note that children can be created long after a span has completed.
+
+  virtual void report(const Span& span) = 0;
```

Comment:
This should probably accept an r-value reference now since the observer can't addRef anymore to keep a copy of the data.

Reply:
The Jaeger observer doesn't need to make a copy, as it serializes the data immediately. I'd guess other observers would do the same.

The problem with taking an rvalue here is that a multiplexer implementation would be forced to make copies for each destination observer, even if the destination observer doesn't actually need its own copy.

That said, if we discover later that some observer implementation _does_ need to keep a copy, we can always revisit this.

Reply:
It's more for the case where we want to allow the span to be processed outside of the destruction of the builder. For example, serializing the span to a Jaeger span arguably shouldn't happen when the builder is trying to deconstruct, it should happen later in the event loop when we're not processing anything, but to do that we'd need to move the span data to the promised function that will be evaluated later.

As for a multiplexing observer, I'd think the observer would immediately want to move the span data on the heap, perhaps into itself, so it can pass a reference counted view to all the other observers (which would use a different API) that might need to hold the data longer than the initial call.

Though ultimately you're right we can just switch it in the future so it doesn't necessarily matter at this moment.

Reply:
I don't think serializing the span synchronously is a problem. It's not that big. I'd expect the machinery to do it asynchronously would cost more than it's worth.

My hope is that the multiplexing observer can stack on top of observers that implement the same `SpanObserver` interface, since we wouldn't want to have to implement each underlying interface twice.

But yeah, I think it's best to defer the debate until we actually need to implement it and know what the requirements are.